src/s3/multipartcopy.rs

```diff
@@ -0,0 +1,311 @@
+use aws_sdk_s3::error::SdkError;
+use aws_sdk_s3::operation::complete_multipart_upload::CompleteMultipartUploadError;
+use aws_sdk_s3::operation::create_multipart_upload::{
+    CreateMultipartUploadError, CreateMultipartUploadOutput,
+};
+use aws_sdk_s3::operation::head_object::HeadObjectError;
+use aws_sdk_s3::types::{CompletedMultipartUpload, CompletedPart};
+use aws_sdk_s3::Client;
+use bytesize::MIB;
+use conv::{prelude::*, FloatError, RangeError};
+use futures::stream::{self, StreamExt, TryStreamExt};
+use std::error::Error as StdError;
+use std::num::TryFromIntError;
+use std::sync::Arc;
+use thiserror::Error;
+use tracing::instrument;
+use typed_builder::TypedBuilder;
+
+use super::S3Object;
+
+/// Retrieves the size of the source object in the specified S3 bucket.
+///
+/// # Arguments
+///
+/// * `client` - A reference to the S3 client.
+/// * `bucket` - The name of the S3 bucket.
+/// * `key` - The key of the source object.
+///
+/// # Returns
+///
+/// The size of the source object in bytes.
+#[instrument(skip(client))]
+pub async fn get_source_size(
+    client: &Client,
+    bucket: &str,
+    key: &str,
+) -> Result<i64, aws_sdk_s3::error::SdkError<HeadObjectError>> {
+    let head_object = client.head_object().bucket(bucket).key(key).send().await?;
+
+    Ok(head_object.content_length().unwrap_or(0))
+}
+
+const DEFAULT_COPY_PART_SIZE: i64 = 50 * MIB as i64;
```

Comment:
**question;** Is the choice of 50MB here significant? Could we leave a small comment justifying this choice. If it's mostly arbitrary then we can just mention that 👍